File: backend/internal/core/dto/usermedication.go
package dto

import (
	"backend/internal/core/shared"
	"time"

	"github.com/google/uuid"
)

type IntakeSchedule struct {
	TimeSlot   shared.TimeSlot `json:"time_slot"   validate:"required,oneof=morning noon evening night"`
	DoseAmount float64         `json:"dose_amount" validate:"required,gt=0"`
}

type UserMedicationCreateRequest struct {
	MedicationID uuid.UUID        `json:"medication_id" validate:"required"`
	BoxesOwned   int              `json:"boxes_owned"   validate:"required,min=1"`
	Schedules    []IntakeSchedule `json:"schedules"     validate:"required,min=1,dive"`
	DurationDays int              `json:"duration_days" validate:"required,min=1"`
}

type UserMedicationUpdateRequest struct {
	BoxesOwned *int              `json:"boxes_owned,omitempty" validate:"omitempty,min=1"`
	Schedules  *[]IntakeSchedule `json:"schedules,omitempty"   validate:"omitempty,min=1,dive"`
	Active     *bool             `json:"active,omitempty"`
}

type UserMedicationResponse struct {
	ID           uuid.UUID        `json:"id"`
	UserID       uuid.UUID        `json:"user_id"`
	MedicationID uuid.UUID        `json:"medication_id"`
	BoxesOwned   int              `json:"boxes_owned"`
	Schedules    []IntakeSchedule `json:"schedules"`
	DurationDays int              `json:"duration_days"`
	StartAt      time.Time        `json:"start_at"`
	Active       bool             `json:"active"`
	CreatedAt    time.Time        `json:"created_at"`
}

type UserMedicationStatsResponse struct {
	TotalPills             int       `json:"total_pills"`
	UsedPills              float64   `json:"used_pills"`
	RemainingPills         float64   `json:"remaining_pills"`
	DailyConsumption       float64   `json:"daily_consumption"`
	EstimatedDaysRemaining int       `json:"estimated_days_remaining"`
	EstimatedEndDate       time.Time `json:"estimated_end_date"`
	PlannedDurationDays    int       `json:"planned_duration_days"`
	DaysElapsed            int       `json:"days_elapsed"`
	PlannedDaysRemaining   int       `json:"planned_days_remaining"`
	WarningLevel           string    `json:"warning_level"` // "normal", "warning", "critical"
}

File: mobile/lib/data/models/user_medication.dart
import 'medication.dart';

class MedicationSchedule {
  final String timeSlot; // "morning", "noon", "evening", "night"
  final double doseAmount;

  MedicationSchedule({
    required this.timeSlot,
    required this.doseAmount,
  });

  factory MedicationSchedule.fromJson(Map<String, dynamic> json) {
    return MedicationSchedule(
      timeSlot: json['time_slot'] ?? '',
      doseAmount: (json['dose_amount'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'time_slot': timeSlot,
      'dose_amount': doseAmount,
    };
  }
}

class UserMedication {
  final String id;
  final String userId;
  final String medicationId;
  final int boxesOwned;
  final List<MedicationSchedule> schedules;
  final int durationDays;
  final String startAt;
  final bool active;
  final String createdAt;
  final Medication? medication;

  UserMedication({
    required this.id,
    required this.userId,
    required this.medicationId,
    required this.boxesOwned,
    required this.schedules,
    required this.durationDays,
    required this.startAt,
    required this.active,
    required this.createdAt,
    this.medication,
  });

  factory UserMedication.fromJson(Map<String, dynamic> json) {
    return UserMedication(
      id: json['id'] ?? '',
      userId: json['user_id'] ?? '',
      medicationId: json['medication_id'] ?? '',
      boxesOwned: json['boxes_owned'] ?? 0,
      schedules: json['schedules'] != null
          ? (json['schedules'] as List)
              .map((s) => MedicationSchedule.fromJson(s))
              .toList()
          : [],
      durationDays: json['duration_days'] ?? 0,
      startAt: json['start_at'] ?? '',
      active: json['active'] ?? false,
      createdAt: json['created_at'] ?? '',
      medication: json['medication'] != null
          ? Medication.fromJson(json['medication'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'medication_id': medicationId,
      'boxes_owned': boxesOwned,
      'schedules': schedules.map((s) => s.toJson()).toList(),
      'duration_days': durationDays,
      'start_at': startAt,
      'active': active,
      'created_at': createdAt,
      if (medication != null) 'medication': medication!.toJson(),
    };
  }
}

class UserMedicationCreateRequest {
  final String medicationId;
  final int boxesOwned;
  final List<MedicationSchedule> schedules;
  final int durationDays;

  UserMedicationCreateRequest({
    required this.medicationId,
    required this.boxesOwned,
    required this.schedules,
    required this.durationDays,
  });

  Map<String, dynamic> toJson() {
    return {
      'medication_id': medicationId,
      'boxes_owned': boxesOwned,
      'schedules': schedules.map((s) => s.toJson()).toList(),
      'duration_days': durationDays,
    };
  }
}
